statefulset/scale_test.go

```diff
@@ -0,0 +1,227 @@
+package statefulset
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestScaledown(t *testing.T) {
+	type state struct {
+		replicas          int32
+		annotationState   string
+		annotationScaleUp string
+		annotationReplica string
+		statusReplicas    int32
+	}
+	type tCase struct {
+		in   state
+		out  state
+		done bool
+	}
+	tcs := map[string]tCase{
+		"should scale down and wait": {
+			in: state{
+				replicas:       6,
+				statusReplicas: 5,
+			},
+			out: state{
+				replicas:          0,
+				statusReplicas:    5,
+				annotationReplica: "6",
+			},
+			done: false,
+		},
+		"should scale down and wait, even if zero replicas running": {
+			in: state{
+				replicas:       2,
+				statusReplicas: 0,
+			},
+			out: state{
+				replicas:          0,
+				statusReplicas:    0,
+				annotationReplica: "2",
+			},
+			done: false,
+		},
+		"should keep waiting": {
+			in: state{
+				replicas:          0,
+				statusReplicas:    2,
+				annotationReplica: "4",
+			},
+			out: state{
+				replicas:          0,
+				statusReplicas:    2,
+				annotationReplica: "4",
+			},
+			done: false,
+		},
+		"should proceed": {
+			in: state{
+				replicas:          0,
+				statusReplicas:    0,
+				annotationReplica: "4",
+			},
+			out: state{
+				replicas:          0,
+				statusReplicas:    0,
+				annotationReplica: "4",
+			},
+			done: true,
+		},
+		"should not scale down if scalUp job is running": {
+			in: state{
+				replicas:          2,
+				statusReplicas:    1,
+				annotationScaleUp: "true",
+				annotationReplica: "2",
+			},
+			out: state{
+				replicas:          2,
+				statusReplicas:    1,
+				annotationReplica: "2",
+			},
+			done: true,
+		},
+	}
+	for k, tc := range tcs {
+		t.Run(k, func(t *testing.T) {
+			assert := assert.New(t)
+			require := require.New(t)
+			sts := appsv1.StatefulSet{
```

Comment:
This could be moved to a helper function too. There are some other places moving to a helper could be applied too.